File: src/main/java/com/hy/frame/view/GrayImageView.java
package com.hy.frame.view;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.ColorMatrix;
import android.graphics.ColorMatrixColorFilter;
import android.graphics.Paint;
import android.util.AttributeSet;

/**
 * 可变灰色图
 * author HeYan
 * time 2015/12/31 10:06
 */
public class GrayImageView extends android.support.v7.widget.AppCompatImageView {
    public GrayImageView(Context context) {
        super(context);
    }

    public GrayImageView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public GrayImageView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    private boolean gray;

    public boolean isGray() {
        return gray;
    }

    public void setGray(boolean gray) {
        this.gray = gray;
    }

    @Override
    public void setImageBitmap(Bitmap bm) {
        if (bm == null) return;
        if (gray) {
            int width = bm.getWidth();
            int height = bm.getHeight();
            Bitmap grayBmp = Bitmap.createBitmap(width, height, Bitmap.Config.RGB_565);
            Canvas c = new Canvas(grayBmp);
            Paint paint = new Paint();
            ColorMatrix cm = new ColorMatrix();
            cm.setSaturation(0);
            ColorMatrixColorFilter f = new ColorMatrixColorFilter(cm);
            paint.setColorFilter(f);
            c.drawBitmap(bm, 0, 0, paint);
            super.setImageBitmap(grayBmp);
        } else
            super.setImageBitmap(bm);
    }
}

File: src/main/java/com/hy/frame/view/GrayImageView.kt
package com.hy.frame.view

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.ColorMatrix
import android.graphics.ColorMatrixColorFilter
import android.graphics.Paint
import android.util.AttributeSet

/**
 * 可变灰色图
 * author HeYan
 * time 2015/12/31 10:06
 */
class GrayImageView : android.support.v7.widget.AppCompatImageView {
    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr)

    var isGray: Boolean = false

    override fun setImageBitmap(bm: Bitmap?) {
        if (bm == null) return
        if (isGray) {
            val width = bm.width
            val height = bm.height
            val grayBmp = Bitmap.createBitmap(width, height, Bitmap.Config.RGB_565)
            val c = Canvas(grayBmp)
            val paint = Paint()
            val cm = ColorMatrix()
            cm.setSaturation(0f)
            val f = ColorMatrixColorFilter(cm)
            paint.colorFilter = f
            c.drawBitmap(bm, 0f, 0f, paint)
            super.setImageBitmap(grayBmp)
        } else
            super.setImageBitmap(bm)
    }
}
